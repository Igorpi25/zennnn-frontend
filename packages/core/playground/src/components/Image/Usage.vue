<template>
  <Image
    lazy-src="https://picsum.photos/id/11/10/6"
    max-height="150"
    max-width="250"
    src="https://picsum.photos/id/11/500/300"
  ></Image>
</template>
