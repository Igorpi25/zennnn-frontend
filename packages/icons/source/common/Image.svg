<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" fill="none"><path d="M9.828 8.219a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0z" fill="#676767"/><path d="M18 4H6c-1.103 0-2 .897-2 2v12c0 1.103.897 2 2 2h12c1.103 0 2-.897 2-2V6c0-1.103-.897-2-2-2zM6 5.333h12c.368 0 .667.3.667.667v5.724L16.47 9.529a.667.667 0 00-.942 0l-4.696 4.695-1.695-1.695a.667.667 0 00-.943 0L5.333 15.39V6c0-.368.3-.667.667-.667z" fill="#676767"/></svg>