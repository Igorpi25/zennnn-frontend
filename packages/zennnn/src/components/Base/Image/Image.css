@layer components {
  .image {
    flex: 1 0 auto;
    @apply z-0;
    @apply flex relative overflow-hidden max-w-full;
  }

  .image__sizer {
    flex: 1 0 0px;
    @apply transition-padding-bottom duration-200 ease-out-quart;
  }

  .image__content {
    flex: 1 0 0px;
    @apply max-w-full;
  }

  .image__sizer ~ .image__content {
    margin-left: -100%;
  }

  .image__image,
  .image__placeholder {
    z-index: -1;
    @apply absolute top-0 left-0 w-full h-full;
  }

  .image__image {
    @apply bg-no-repeat;
  }
  .image__image--preload {
    filter: blur(2px);
  }
  .image__image--contain {
    @apply bg-contain;
  }
  .image__image--cover {
    @apply bg-cover;
  }
}
