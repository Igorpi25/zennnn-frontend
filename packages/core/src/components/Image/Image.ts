import {
  h,
  ref,
  watch,
  reactive,
  computed,
  onBeforeMount,
  nextTick,
  vShow,
  Transition,
  withDirectives,
  defineComponent,
  VNode,
  PropType,
  TransitionProps,
  mergeProps,
} from 'vue'

import { useAspectRatioProps, useAspectRatio, useDimensionProps, useDimension, Intersect } from 'vue-supp'

import './Image.css'

const hasIntersect = typeof window !== 'undefined' && 'IntersectionObserver' in window

export interface srcObject {
  src?: string
  srcset?: string
  lazySrc?: string
  aspect: number
}

export default defineComponent({
  name: 'Image',

  props: {
    ...useAspectRatioProps(),
    ...useDimensionProps(),
    alt: String,
    cover: Boolean,
    eager: Boolean,
    lazySrc: String,
    options: {
      type: Object as PropType<IntersectionObserverInit>,
      // For more information on types, navigate to:
      // https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API
      default: () => ({
        root: undefined,
        rootMargin: undefined,
        threshold: undefined,
      }),
    },
    position: {
      type: String,
      default: 'center center',
    },
    sizes: String,
    src: {
      type: [String, Object] as PropType<string | srcObject>,
      default: '',
    },
    srcset: String,
    transition: {
      type: [Boolean, String] as PropType<string | false>,
      default: 'fade-transition',
    },
  },

  emits: ['load', 'error', 'loadstart'],

  setup (props, { slots, emit }) {
    const currentSrc = ref<string>('')
    const image = ref<HTMLImageElement>()
    const state = ref<'idle' | 'loading' | 'loaded' | 'error'>('idle')
    const naturalWidth = ref<number>()
    const naturalHeight = ref<number>()

    const { dimensionStyles } = useDimension(props)

    const aspectRatio = computed(() => normalisedSrc.value.aspect || naturalWidth.value! / naturalHeight.value! || 0)

    const aspectRatioProps = reactive({
      aspectRatio,
    })
    const { aspectStyle } = useAspectRatio(aspectRatioProps)

    const normalisedSrc = computed(() => {
      return props.src && typeof props.src === 'object'
        ? {
            src: props.src.src,
            srcset: props.srcset || props.src.srcset,
            lazySrc: props.lazySrc || props.src.lazySrc,
            aspect: Number(props.aspectRatio || props.src.aspect),
          }
        : {
            src: props.src,
            srcset: props.srcset,
            lazySrc: props.lazySrc,
            aspect: Number(props.aspectRatio || 0),
          }
    })

    const containClasses = computed(() => ({
      'bg-cover': props.cover,
      'bg-contain': !props.cover,
    }))

    const __image = computed(() => {
      if (!normalisedSrc.value.src || state.value === 'idle') return

      const img = h('img', {
        ref: image,
        class: ['image__image', containClasses.value],
        src: normalisedSrc.value.src,
        srcset: normalisedSrc.value.srcset,
        sizes: props.sizes,
        onLoad,
        onError,
      })

      const sources = slots.sources?.()

      return genTransition(
        { appear: true },
        withDirectives(
          sources
            ? h('picture', { class: 'image__picture' }, [sources, img])
            : img,
          [[vShow, state.value === 'loaded']],
        ),
      )
    })

    const __preloadImage = computed(() => {
      const img = h('img', {
        src: normalisedSrc.value.lazySrc,
        class: ['image__image image__image--preload', containClasses.value],
      })
      return genTransition(
        {},
        normalisedSrc.value.lazySrc && state.value !== 'loaded' ? img : undefined,
      )
    })

    const __placeholder = computed(() => {
      if (!slots.placeholder) return

      const placeholder = state.value === 'loading' || (state.value === 'error' && !slots.error)
        ? h('div', { class: 'image__placeholder' }, slots.placeholder())
        : undefined

      return genTransition({ appear: true }, placeholder)
    })

    const __error = computed(() => {
      if (!slots.error) return

      const error = state.value === 'error'
        ? h('div', { class: 'image__error' }, slots.error())
        : undefined

      return genTransition({ appear: true }, error)
    })

    watch(() => props.src, () => {
      init(undefined, undefined, state.value !== 'idle')
    })
    // TODO: getSrc when window width changes

    onBeforeMount(() => init())

    const init = (entries?: IntersectionObserverEntry[], observer?: IntersectionObserver, isIntersecting?: boolean) => {
      // If the current browser supports the intersection
      // observer api, the image is not observable, and
      // the eager prop isn't being used, do not load
      if (
        hasIntersect &&
        !isIntersecting &&
        !props.eager
      ) return

      state.value = 'loading'
      nextTick(() => {
        emit('loadstart', image.value?.currentSrc || normalisedSrc.value.src)
        if (!aspectRatio.value) pollForSize(image.value!)
        getSrc()
      })

      if (normalisedSrc.value.lazySrc) {
        const lazyImg = new Image()
        lazyImg.src = normalisedSrc.value.lazySrc
        pollForSize(lazyImg, null)
      }
    }

    const onLoad = () => {
      getSrc()
      state.value = 'loaded'
      emit('load', image.value?.currentSrc || normalisedSrc.value.src)
    }

    const onError = () => {
      state.value = 'error'
      emit('error', image.value?.currentSrc || normalisedSrc.value.src)
    }

    const getSrc = () => {
      if (image.value) currentSrc.value = image.value.currentSrc || image.value.src
    }

    const pollForSize = (img: HTMLImageElement, timeout: number | null = 100) => {
      const poll = () => {
        const { naturalHeight: imgHeight, naturalWidth: imgWidth } = img

        if (imgHeight || imgWidth) {
          naturalWidth.value = imgWidth
          naturalHeight.value = imgHeight
        } else if (!img.complete && state.value === 'loading' && timeout != null) {
          setTimeout(poll, timeout)
        } else if (img.currentSrc.endsWith('.svg') || img.currentSrc.startsWith('data:image/svg+xml')) {
          naturalWidth.value = 1
          naturalHeight.value = 1
        }
      }

      poll()
    }

    const genTransition = (data: TransitionProps, vNodes: VNode | VNode[] | undefined) => {
      if (!props.transition) return vNodes

      return h(
        Transition,
        mergeProps({ name: props.transition }, data as any) as TransitionProps,
        () => vNodes,
      )
    }

    const genContent = () => {
      return h('div', {
        class: 'image__content',
      }, slots.default?.())
    }

    const genDirectives = (content: VNode) => {
      if (!hasIntersect) return content
      return withDirectives(
        content,
        [
          [Intersect, { handler: init, options: props.options }, '', { once: true }],
        ],
      )
    }

    return () => {
      const data = {
        class: 'image',
        style: dimensionStyles.value,
        'aria-label': props.alt,
        role: props.alt ? 'img' : undefined,
      }

      const sizer = h('div', {
        style: aspectStyle.value,
        class: 'image__sizer',
      })

      const children = [
        sizer,
        __image.value,
        __preloadImage.value,
        __placeholder.value,
        __error.value,
        genContent(),
      ]

      return genDirectives(h('div', data, children))
    }
  },
})
