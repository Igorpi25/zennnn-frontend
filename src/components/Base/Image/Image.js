import {
  h,
  ref,
  watch,
  reactive,
  computed,
  onMounted,
  Transition,
  withDirectives,
} from 'vue'

import './Image.css'

import { useAspectRatioProps, useAspectRatio, dimensions, Intersect } from 'uipart'

const { useDimensionsProps, useDimensions } = dimensions()

const hasIntersect = typeof window !== 'undefined' && 'IntersectionObserver' in window

export default {
  name: 'Image',

  props: {
    ...useAspectRatioProps(),
    ...useDimensionsProps(),
    alt: String,
    contain: Boolean,
    eager: Boolean,
    gradient: String,
    lazySrc: String,
    options: {
      type: Object,
      // For more information on types, navigate to:
      // https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API
      default: () => ({
        root: undefined,
        rootMargin: undefined,
        threshold: undefined,
      }),
    },
    position: {
      type: String,
      default: 'center center',
    },
    sizes: String,
    src: {
      type: [String, Object],
      default: '',
    },
    srcset: String,
    transition: {
      type: [Boolean, String],
      default: 'fade-transition',
    },
  },

  emits: ['load', 'error'],

  setup (props, { slots, emit }) {
    // Data
    const currentSrc = ref('')
    const image = ref(null)
    const isLoading = ref(true)
    const calculatedAspectRatio = ref(undefined)
    const naturalWidth = ref(undefined)
    const hasError = ref(false)

    const measureProps = reactive({
      height: computed(() => props.height),
      maxHeight: computed(() => props.maxHeight),
      maxWidth: computed(() => props.maxWidth),
      minHeight: computed(() => props.minHeight),
      minWidth: computed(() => props.minWidth),
      width: computed(() => props.width),
    })
    const { dimensionsStyles } = useDimensions(measureProps)

    const aspectRatioProps = reactive({
      aspectRatio: computed(() => Number(normalisedSrc.value.aspect || calculatedAspectRatio.value)),
    })
    const { aspectStyle } = useAspectRatio(aspectRatioProps)

    const normalisedSrc = computed(() => {
      return props.src && typeof props.src === 'object'
        ? {
          src: props.src.src,
          srcset: props.srcset || props.src.srcset,
          lazySrc: props.lazySrc || props.src.lazySrc,
          aspect: Number(props.aspectRatio || props.src.aspect),
        } : {
          src: props.src,
          srcset: props.srcset,
          lazySrc: props.lazySrc,
          aspect: Number(props.aspectRatio || 0),
        }
    })

    const cachedSizer = computed(() => {
      if (!aspectStyle.value) return undefined

      return h('div', {
        style: aspectStyle.value,
        class: 'image__sizer',
      })
    })

    const cachedImage = computed(() => {
      if (!(normalisedSrc.value.src || normalisedSrc.value.lazySrc || props.gradient)) return []

      const backgroundImage = []
      const src = isLoading.value ? normalisedSrc.value.lazySrc : currentSrc.value

      if (props.gradient) backgroundImage.push(`linear-gradient(${props.gradient})`)
      if (src) backgroundImage.push(`url("${src}")`)

      const image = h('div', {
        class: {
          image__image: true,
          'image__image--preload': isLoading.value,
          'image__image--contain': props.contain,
          'image__image--cover': !props.contain,
        },
        style: {
          backgroundImage: backgroundImage.join(', '),
          backgroundPosition: props.position,
        },
        key: +isLoading.value,
      })

      if (!props.transition) return image

      return h(Transition, {
        name: props.transition,
        mode: 'in-out',
      }, {
        default: () => image,
      })
    })

    // Watch
    watch(() => props.src, () => {
      // Force re-init when src changes
      if (!isLoading.value) init(undefined, undefined, true)
      else loadImage()
    })
    // TODO: add watcher for client width change with 'getSrc' function

    // Hooks
    onMounted(() => {
      init()
    })

    // Methods
    const init = (entries, observer, isIntersecting) => {
      // If the current browser supports the intersection
      // observer api, the image is not observable, and
      // the eager prop isn't being used, do not load
      if (
        hasIntersect &&
        !isIntersecting &&
        !props.eager
      ) return

      if (normalisedSrc.value.lazySrc) {
        const lazyImg = new Image()
        lazyImg.src = normalisedSrc.value.lazySrc
        pollForSize(lazyImg, null)
      }

      if (normalisedSrc.value.src) loadImage()
    }

    const onLoad = () => {
      getSrc()
      isLoading.value = false
      emit('load', props.src)
    }

    const onError = () => {
      hasError.value = true
      emit('error', props.src)
    }

    const getSrc = () => {
      if (image.value) currentSrc.value = image.value.currentSrc || image.value.src
    }

    const loadImage = () => {
      const _image = new Image()
      image.value = _image

      _image.onload = () => {
        if (_image.decode) {
          _image.decode().catch((err) => {
            // eslint-disable-next-line
            console.warn(
              'Failed to decode image, trying to render anyway\n\n' +
              `src: ${normalisedSrc.value.src}` +
              (err.message ? `\nOriginal error: ${err.message}` : ''),
            )
          }).then(onLoad)
        } else {
          onLoad()
        }
      }
      _image.onerror = onError

      hasError.value = false
      _image.src = normalisedSrc.value.src
      props.sizes && (_image.sizes = props.sizes)
      normalisedSrc.value.srcset && (_image.srcset = normalisedSrc.value.srcset)

      props.aspectRatio || pollForSize(_image)
      getSrc()
    }

    const pollForSize = (img, timeout = 100) => {
      const poll = () => {
        const { naturalHeight: _naturalHeight, naturalWidth: _naturalWidth } = img

        if (_naturalHeight || _naturalWidth) {
          naturalWidth.value = _naturalWidth
          calculatedAspectRatio.value = _naturalWidth / _naturalHeight
        } else {
          timeout != null && !hasError.value && setTimeout(poll, timeout)
        }
      }

      poll()
    }

    const genContent = () => {
      const data = {
        class: 'image__content',
      }
      if (naturalWidth.value) {
        data.style = { width: `${naturalWidth.value}px` }
      }

      return h('div', data, slots.default ? slots.default() : undefined)
    }

    const genPlaceholder = () => {
      if (slots.placeholder) {
        const placeholder = isLoading.value
          ? h('div', {
            class: 'image__placeholder',
          }, slots.placeholder())
          : undefined

        if (!props.transition) return placeholder

        return h(Transition, {
          appear: true,
          name: props.transition,
        }, {
          default: () => placeholder,
        })
      }
    }

    const genDirectives = (content) => {
      if (!hasIntersect) return content
      return withDirectives(
        content,
        [
          [Intersect, { handler: init, options: props.options }, null, { once: true }],
        ],
      )
    }

    return {
      aspectStyle,
      dimensionsStyles,
      cachedSizer,
      cachedImage,
      genContent,
      genPlaceholder,
      genDirectives,
    }
  },

  render () {
    const data = {
      class: 'image',
      style: this.dimensionsStyles,
      'aria-label': this.alt,
      role: this.alt ? 'img' : undefined,
    }

    const children = [
      this.cachedSizer,
      this.cachedImage,
      this.genPlaceholder(),
      this.genContent(),
    ]

    return this.genDirectives(h('div', data, children))
  },
}
