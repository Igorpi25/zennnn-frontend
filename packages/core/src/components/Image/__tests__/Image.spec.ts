import { mount, VueWrapper } from '@vue/test-utils'
import { h, nextTick } from 'vue'
import Image from '../Image'

describe('Image.ts', () => {
  const LOAD_FAILURE_SRC = 'LOAD_FAILURE_SRC'
  const LOAD_SUCCESS_SRC = 'LOAD_SUCCESS_SRC'

  let mountFunction: (options?: Record<string, any>) => VueWrapper<any>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(Image, {
        ...options,
        props: {
          eager: true,
          ...options.props,
        },
      })
    }
  })

  beforeAll(() => {
    jest.useFakeTimers()
    Object.defineProperty((global as any).Image.prototype, 'src', {
      get () {
        return this._currentSrc
      },
      set (src) {
        this._currentSrc = src
        if (src === LOAD_FAILURE_SRC) {
          setTimeout(() => {
            this.dispatchEvent(new ErrorEvent('error'))
          })
        } else {
          setTimeout(() => {
            this._naturalWidth = 1600
            this._naturalHeight = 900
            this.dispatchEvent(new Event('load', { bubbles: false }))
          })
        }
      },
    })
    Object.defineProperty((global as any).Image.prototype, 'currentSrc', {
      get () {
        return this._currentSrc
      },
    })
    Object.defineProperty((global as any).Image.prototype, 'naturalWidth', {
      get () { return this._naturalWidth },
    })
    Object.defineProperty((global as any).Image.prototype, 'naturalHeight', {
      get () { return this._naturalHeight },
    })
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  it('should load', async () => {
    const wrapper = mountFunction({
      props: { src: LOAD_SUCCESS_SRC },
    })

    expect(wrapper.html()).toMatchSnapshot()

    jest.runOnlyPendingTimers()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a <picture>', async () => {
    const wrapper = mountFunction({
      props: { src: LOAD_SUCCESS_SRC },
      slots: {
        sources: () => h('source', { srcset: LOAD_SUCCESS_SRC + '.webp' }),
      },
    })

    jest.runOnlyPendingTimers()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should display placeholders', async () => {
    const wrapper = mountFunction({
      props: {
        src: 'full_src',
        lazySrc: 'lazy_src',
      },
      slots: {
        placeholder: () => h('div', ['loading...']),
      },
    })

    expect(wrapper.html()).toMatchSnapshot()

    jest.runOnlyPendingTimers()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should emit errors', () => {
    const error = jest.fn()
    mountFunction({
      props: {
        src: LOAD_FAILURE_SRC,
        onError: error,
      },
    })

    jest.runOnlyPendingTimers()

    expect(error).toHaveBeenCalledTimes(1)
    expect(error).toHaveBeenCalledWith(LOAD_FAILURE_SRC)
  })

  it('should display error slot', async () => {
    const wrapper = mountFunction({
      props: {
        src: LOAD_FAILURE_SRC,
      },
      slots: {
        error: () => h('div', 'Error loading image'),
      },
    })

    jest.runOnlyPendingTimers()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have aria attributes', async () => {
    const wrapper = mountFunction({
      props: {
        src: LOAD_SUCCESS_SRC,
        alt: 'role and aria label',
      },
    })

    jest.runOnlyPendingTimers()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should update src', async () => {
    const wrapper = mountFunction({
      props: {
        src: LOAD_SUCCESS_SRC,
      },
    })

    jest.runOnlyPendingTimers()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ src: `${LOAD_SUCCESS_SRC}-1` })

    jest.runOnlyPendingTimers()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should update src while still loading', async () => {
    const wrapper = mountFunction({
      props: {
        src: LOAD_SUCCESS_SRC,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ src: `${LOAD_SUCCESS_SRC}-1` })
    jest.runOnlyPendingTimers()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })
})
