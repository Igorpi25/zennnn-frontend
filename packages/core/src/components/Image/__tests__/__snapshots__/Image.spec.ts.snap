// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image.ts should display error slot 1`] = `
<div class="image">
  <div class="image__sizer"></div>
  <transition-stub><img class="image__image bg-contain" style="display: none;"></transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <transition-stub>
    <div class="image__error">
      <div>Error loading image</div>
    </div>
  </transition-stub>
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should display placeholders 1`] = `
<div class="image">
  <div class="image__sizer"></div>
  <transition-stub><img class="image__image bg-contain" style="display: none;"></transition-stub>
  <transition-stub><img class="image__image image__image--preload bg-contain"></transition-stub>
  <transition-stub>
    <div class="image__placeholder">
      <div>loading...</div>
    </div>
  </transition-stub>
  <!---->
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should display placeholders 2`] = `
<div class="image">
  <div class="image__sizer" style="padding-bottom: 56.25%;"></div>
  <transition-stub><img class="image__image bg-contain"></transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should have aria attributes 1`] = `
<div class="image" aria-label="role and aria label" role="img">
  <div class="image__sizer" style="padding-bottom: 56.25%;"></div>
  <transition-stub><img class="image__image bg-contain"></transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <!---->
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should load 1`] = `
<div class="image">
  <div class="image__sizer"></div>
  <transition-stub><img class="image__image bg-contain" style="display: none;"></transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <!---->
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should load 2`] = `
<div class="image">
  <div class="image__sizer" style="padding-bottom: 56.25%;"></div>
  <transition-stub><img class="image__image bg-contain"></transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <!---->
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should render a <picture> 1`] = `
<div class="image">
  <div class="image__sizer" style="padding-bottom: 56.25%;"></div>
  <transition-stub>
    <picture class="image__picture">
      <source srcset="LOAD_SUCCESS_SRC.webp"><img class="image__image bg-contain">
    </picture>
  </transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <!---->
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should update src 1`] = `
<div class="image">
  <div class="image__sizer" style="padding-bottom: 56.25%;"></div>
  <transition-stub><img class="image__image bg-contain"></transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <!---->
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should update src 2`] = `
<div class="image">
  <div class="image__sizer" style="padding-bottom: 56.25%;"></div>
  <transition-stub><img class="image__image bg-contain"></transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <!---->
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should update src while still loading 1`] = `
<div class="image">
  <div class="image__sizer"></div>
  <transition-stub><img class="image__image bg-contain" style="display: none;"></transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <!---->
  <div class="image__content"></div>
</div>
`;

exports[`Image.ts should update src while still loading 2`] = `
<div class="image">
  <div class="image__sizer" style="padding-bottom: 56.25%;"></div>
  <transition-stub><img class="image__image bg-contain"></transition-stub>
  <transition-stub>
    <!---->
  </transition-stub>
  <!---->
  <!---->
  <div class="image__content"></div>
</div>
`;
